src/Models/UserModel.php

```diff
@@ -225,12 +225,18 @@ public function activate(User $user): void
      */
     public function insert($data = null, bool $returnID = true)
     {
-        $this->tempUser = $data instanceof User ? $data : null;
+        // Clone User object for not changing the passed object.
+        $this->tempUser = $data instanceof User ? clone $data : null;
 
         $result = parent::insert($data, $returnID);
 
         $this->checkQueryReturn($result);
 
+        // Set user id to the passed User object.
+        if ($data instanceof User) {
+            $data->id = $this->insertID;
+        }
+
```

Comment:
I'm not a fan of modifying the object at all; it gives the false impression of database equivalency. We have no created/updated timestamps, or other fields coming from the database, or effects of stored protocols and field defaults. If we really need the ID on the object instead of just in the return would should reload the whole thing.

Reply:
> We have no created/updated timestamps, or other fields coming from the database, or effects of stored protocols and field defaults.

Exactly.
We should discourage the use of User objects that are not fetched from the database?
We should think the following code is just a misuse?

```php
        $user           = new User();
        $user->username = 'foo';
        $user->email    = 'foo@bar.com';
        $user->password = 'password';
        $user->active   = false;

        $users->save($user);

        $identity = $user->getEmailIdentity();
```

I added the set use id logic because if I remove it, the test fails:
```
There was 1 failure:

1) Tests\Unit\UserModelTest::testSaveNewUserAndGetEmailIdentity
"$user->id" must not be null. in /Users/kenji/work/codeigniter/official/codeigniter-shield/src/Models/UserIdentityModel.php:221

Caused by
AssertionError: "$user->id" must not be null. in /Users/kenji/work/codeigniter/official/codeigniter-shield/src/Models/UserIdentityModel.php:221
Stack trace:
#0 /Users/kenji/work/codeigniter/official/codeigniter-shield/src/Models/UserIdentityModel.php(221): assert(false, '"$user->id" mus...')
#1 /Users/kenji/work/codeigniter/official/codeigniter-shield/src/Entities/User.php(81): CodeIgniter\Shield\Models\UserIdentityModel->getIdentities(Object(CodeIgniter\Shield\Entities\User))
#2 /Users/kenji/work/codeigniter/official/codeigniter-shield/src/Entities/User.php(95): CodeIgniter\Shield\Entities\User->populateIdentities()
#3 /Users/kenji/work/codeigniter/official/codeigniter-shield/src/Entities/User.php(66): CodeIgniter\Shield\Entities\User->getIdentities('email_password')
#4 /Users/kenji/work/codeigniter/official/codeigniter-shield/src/Entities/User.php(134): CodeIgniter\Shield\Entities\User->getIdentity('email_password')
#5 /Users/kenji/work/codeigniter/official/codeigniter-shield/tests/Unit/UserModelTest.php(75): CodeIgniter\Shield\Entities\User->getEmailIdentity()
#6 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/Framework/TestCase.php(1545): Tests\Unit\UserModelTest->testSaveNewUserAndGetEmailIdentity()
#7 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/Framework/TestCase.php(1151): PHPUnit\Framework\TestCase->runTest()
#8 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/Framework/TestResult.php(726): PHPUnit\Framework\TestCase->runBare()
#9 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/Framework/TestCase.php(903): PHPUnit\Framework\TestResult->run(Object(Tests\Unit\UserModelTest))
#10 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/Framework/TestSuite.php(672): PHPUnit\Framework\TestCase->run(Object(PHPUnit\Framework\TestResult))
#11 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/Framework/TestSuite.php(672): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#12 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/Framework/TestSuite.php(672): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#13 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(673): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#14 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/TextUI/Command.php(143): PHPUnit\TextUI\TestRunner->run(Object(PHPUnit\Framework\TestSuite), Array, Array, true)
#15 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/src/TextUI/Command.php(96): PHPUnit\TextUI\Command->run(Array, true)
#16 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/phpunit/phpunit/phpunit(98): PHPUnit\TextUI\Command::main()
#17 /Users/kenji/work/codeigniter/official/codeigniter-shield/vendor/bin/phpunit(123): include('/Users/kenji/wo...')
#18 {main}
```

Reply:
I removed the logic.